command/agent/auth/auth.go

```diff
@@ -169,12 +174,38 @@ func (ah *AuthHandler) Run(ctx context.Context, am AuthMethod) error {
 			}
 		}
 
-		secret, err := clientToUse.Logical().Write(path, data)
-		// Check errors/sanity
-		if err != nil {
-			ah.logger.Error("error authenticating", "error", err, "backoff", backoff.Seconds())
-			backoffOrQuit(ctx, backoff)
-			continue
+		var secret *api.Secret = new(api.Secret)
+		if first && ah.token != "" {
```

Comment:
Do we want to do this before `am.Authenticate(ctx, ah.client)`?

This change also makes me wonder whether we should disable sink writing back the token if it's a token that is being preloaded.

Reply:
> Do we want to do this before am.Authenticate(ctx, ah.client)?

I don't think so, actually. `am.Authenticate` will catch misconfigurations of auto-auth in which case we'll want to bail. Thoughts?

> This change also makes me wonder whether we should disable sink writing back the token if it's a token that is being preloaded.

The sink should still be shared, even if its a preloaded token, because it will need to be rewrapped. Additionally the sink is used in acceptance testing and is a good validation that the token flowed through auto-auth. Thoughts?

Reply:
For the case of approle auto-auth, if the token is being pre-loaded why would we want to check whether the role ID file exists first (within `Authenticate`)? If we're preloading the token, we can probably use that and then fallback to the auto-auth's method of getting the token if that fails.

General config parsing and validation should already be done by the `config` package.

> The sink should still be shared, even if its a preloaded token, because it will need to be rewrapped.

IIRC the token that the sink drops does not have to necessarily be wrapped (i.e. the raw token is written), and it's [off by default](https://www.vaultproject.io/docs/agent/autoauth#wrap_ttl) unless explicitly set to a non-empty value.

Reply:
> For the case of approle auto-auth, if the token is being pre-loaded why would we want to check whether the role ID file exists first (within Authenticate)? If we're preloading the token, we can probably use that and then fallback to the auto-auth's method of getting the token if that fails.

We want to verify its OK because we expect at some point the preloaded token is no longer usable (max ttl) and we will then need to switch to auto-auth. We should ensure that the everything is configured correctly before that happens to avoid errors that may not appear for a long time.

> IIRC the token that the sink drops does not have to necessarily be wrapped (i.e. the raw token is written), and it's off by default unless explicitly set to a non-empty value.

That's true but I still don't see the advantage of disabling it. Less code/special cases to handle and the existing behavior doesn't change, even though where the token came from did.

Reply:
> We should ensure that the everything is configured correctly before that happens to avoid errors that may not appear for a long time.

This is true, though this doesn't mitigate the case where the data returned by `Authenticate` will be out-of-date for the `Logical().Write` call below, when that gets consumed. I'm just slightly concerned by the gap between `Authenticate` and the actual call to auth since this sit in between (same point as you brought up, but from a different angle).

> That's true but I still don't see the advantage of disabling it.

I was thinking about this from a security perspective, since you're now threading/dropping a token that was not necessarily auth'ed by this agent instance. You're right though -- sink file access remains unchanged and that's transparent to apps that consume it, so probably not a big deal if we write the token to sinks.

Reply:
Made some changes and works as expected. 👍